### ISSUE TITLE
Module not found: Error: can't resolve... transitive dependencies

### ISSUE THREAD

**Issue by johnnyreilly** [oven-sh/bun#2257]
### What version of Bun is running?

0.5.7

### What platform is your computer?

It works on my Ubuntu 22 laptop: Linux 5.19.0-32-generic x86_64 x86_64 - it fails on GitHub Actions which is also Ubuntu 22 

### What steps can reproduce the bug?

I'm not sure if this is related, but I've been experiencing:

> Module not found: Error: Can't resolve 'react-loadable' in '/home/runner/work/blog.johnnyreilly.com/blog.johnnyreilly.com/blog-website/node_modules/@docusaurus/core/lib/client/exports'

when I tried to port my Docusaurus blog to Bun.

### What is the expected behavior?

The module (in this case `react-loadable` - but presumably this is more related to transitive dependencies?) to be available.

### What do you see instead?

> Module not found: Error: Can't resolve 'react-loadable' in '/home/runner/work/blog.johnnyreilly.com/blog.johnnyreilly.com/blog-website/node_modules/@docusaurus/core/lib/client/exports'

### Additional information

It works when I run locally on my Ubuntu machine.  I've `rm -rf node_modules` there.  It fails reliably on GitHub Actions: 

See PR: https://github.com/johnnyreilly/blog.johnnyreilly.com/pull/447
See failing action: https://github.com/johnnyreilly/blog.johnnyreilly.com/actions/runs/4308209675/jobs/7514209124

**Comment by Jarred-Sumner** [comment]
I don't think it'll fix it, but could you try with the canary build of bun?

**Comment by Jarred-Sumner** [comment]
@alexlamsl can you take a look?

**Comment by johnnyreilly** [comment]
> I don't think it'll fix it, but could you try with the canary build of bun?

Trying it here: https://github.com/johnnyreilly/blog.johnnyreilly.com/actions/runs/4311115695/jobs/7520202987

No dice.

The funny thing from my perspective, is that my laptop is Ubuntu 22, and by the looks of it GitHub Actions are also. So I'd expect similar behaviour.

But no: it works on my machine; it doesn't in GitHub Actions

**Comment by alexlamsl** [comment]
@johnnyreilly I was able to replicate your GitHub Actions job failure locally &minus; looks like the culprit is the `bun.lockb` which stores the dependency tree information, and seems to be generated by an earlier (read: buggy) version of `bun`. (Specifically, the bug is with dependency aliasing.)

In the presence of `bun.lockb`, `bun install` will parse and reuse the stored structure for `node_modules` placement, thus gives the same invalid result even using the latest version of `bun`.

The solution is simply remove `bun.lockb` and run `bun install` again to regenerate the file again, and update your PR with that.

**Comment by johnnyreilly** [comment]
> and seems to be generated by an earlier (read: buggy) version of bun. (Specifically, the bug is with dependency aliasing.)

I generated bun.lockb with 0.5.7 of Bun - so I'm not sure that's the issue. I'm guessing you're referring to this? https://github.com/oven-sh/bun/pull/1837

Let me regenerate anyway as suggested and let's see what happens.

**Comment by johnnyreilly** [comment]
Regenerated bun.lockb with 0.5.7 of Bun - issue still exists: https://github.com/johnnyreilly/blog.johnnyreilly.com/actions/runs/4320618414/jobs/7541025370

Let me try removing bun.lockb entirely.

**Comment by Jarred-Sumner** [comment]
Try `rm bun.lockb && bun upgrade --canary && bun install`

**Comment by johnnyreilly** [comment]
Losing the bun.lockb works: https://github.com/johnnyreilly/blog.johnnyreilly.com/actions/runs/4320645521/jobs/7541078917

Trying `rm bun.lockb && bun upgrade --canary && bun install` now: https://github.com/johnnyreilly/blog.johnnyreilly.com/actions/runs/4320690231/jobs/7541165397

**Comment by johnnyreilly** [comment]
No dice: https://github.com/johnnyreilly/blog.johnnyreilly.com/actions/runs/4320690231/jobs/7541165397

> Module not found: Error: Can't resolve 'react-loadable' in '/home/runner/work/blog.johnnyreilly.com/blog.johnnyreilly.com/blog-website/node_modules/@docusaurus/core/lib/client/exports'

**Comment by Jarred-Sumner** [comment]
maybe we are not normalizing file paths correctly across projects in the lockfile

maybe it is storing absolute file paths somewhere

**Comment by johnnyreilly** [comment]
It's interesting that running locally it just works, but when run in GHA it fails.  Particularly given that both locally and GHA are running a variant of Ubuntu 22.

**Comment by Jarred-Sumner** [comment]
if you replicate the same number of directory tree levels in both does it work?

like

/foo/bar/baz/beep/app
/literally/any/thing/boop/app

**Comment by johnnyreilly** [comment]
So in GHA it's running here: 

`/home/runner/work/blog.johnnyreilly.com/blog.johnnyreilly.com`

On my machine it's running here:

`/home/john/code/github/blog.johnnyreilly.com`

So quite by chance, it seems to already be the same number of directory tree levels

**Comment by Jarred-Sumner** [comment]
If you do bun install --force, does it work? That forces it to reinstall everything even if the files already exist and disables some caching

**Comment by johnnyreilly** [comment]
tested here: https://github.com/johnnyreilly/blog.johnnyreilly.com/actions/runs/4320936428/jobs/7541659214

A different error:

```
$ cross-env WEBPACK_URL_LOADER_LIMIT=0 docusaurus build
file:///home/runner/work/blog.johnnyreilly.com/blog.johnnyreilly.com/blog-website/node_modules/@docusaurus/core/bin/docusaurus.mjs:12
import cli from 'commander';
       ^^^
SyntaxError: The requested module 'commander' does not provide an export named 'default'
    at ModuleJob._instantiate (node:internal/modules/esm/module_job:124:21)
    at async ModuleJob.run (node:internal/modules/esm/module_job:190:5)

Node.js v18.14.2
error: script "build" exited with code 1 (SIGHUP)
```

Not sure why Node 18 is mentioned. This is `docusaurus.mjs`:

```
#!/usr/bin/env node
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

// @ts-check

import logger from '@docusaurus/logger';
import cli from 'commander';
```

Maybe the `#!/usr/bin/env node` means Node is getting invoked.

Other interesting piece of information: without the `bun.lockb` present [`bun run build` runs just fine and takes 5 mins 45 secs](https://github.com/johnnyreilly/blog.johnnyreilly.com/actions/runs/4320645521)

Run with Node, [`yarn run build` would typically take 6 mins 45 secs](https://github.com/johnnyreilly/blog.johnnyreilly.com/actions/runs/4305480138)

So we know that Bun is faster than Node for the Docusaurus use case.  I'll confess ignorance as to whether [`docusaurus build`](https://docusaurus.io/docs/cli#docusaurus-build-sitedir) is invoking Node anywhere and for how much.

**Comment by Jarred-Sumner** [comment]
I don't think bun is able to run `webpack` yet due to lack of `node:vm` support and possibly a transpiler bug. So those timings are the gap between `bun install` and `yarn install`, except it's also downloading & resolving all the dependencies during `bun install`. This makes sense because docusaurus has a lot of dependencies.

This bug with bun.lockb is interesting and we need to investigate it some more. It's probably another aliasing + hoisting related bug.

**Comment by alexlamsl** [comment]
#2397 may have addressed this &minus; please give it a go when it's available on `bun upgrade --canary`

**Comment by squishyu** [comment]
I'm facing the same issue while trying to build Docusaurus with Bun as the package manager inside a Docker container based on Linux Alpine. It builds fine on my Arch Linux.

I created a [reproduction repo with a Dockerfile](https://github.com/squishyu/bun-docusaurus-react-loadable-bug).

It fails on the `react-loadable` package.

It might be helpful to note that `react-loadable` is listed in the package.json of the `@docusaurus/core` package under an alias:

```json

    "react-loadable": "npm:@docusaurus/react-loadable@5.5.2",

```

If I copy the dependency to the root package.json, it works.

**Comment by haraldvinje** [comment]
I've faced this issue as well when building in Github Actions on `ubuntu-latest`. I use Garuda Linux (arch based) locally and never had the issue there. If I delete the Github Actions cache it builds as expected.

[https://github.com/haraldvinje/tekblogg/actions/runs/6292089078/job/17081192160](https://github.com/haraldvinje/tekblogg/actions/runs/6292089078/job/17081192160
)
```
- info Compiled successfully
- info Linting and checking validity of types...
- info Collecting page data...
unhandledRejection Error: Cannot find module 'next/dist/shared/lib/html-context.js'
Require stack:
- /home/runner/work/tekblogg/tekblogg/tekblogg-web/.next/server/pages/_document.js
- /home/runner/work/tekblogg/tekblogg/tekblogg-web/node_modules/next-sanity-image/node_modules/next/dist/server/require.js
- /home/runner/work/tekblogg/tekblogg/tekblogg-web/node_modules/next-sanity-image/node_modules/next/dist/server/load-components.js
- /home/runner/work/tekblogg/tekblogg/tekblogg-web/node_modules/next-sanity-image/node_modules/next/dist/build/utils.js
- /home/runner/work/tekblogg/tekblogg/tekblogg-web/node_modules/next-sanity-image/node_modules/next/dist/build/worker.js
- /home/runner/work/tekblogg/tekblogg/tekblogg-web/node_modules/next-sanity-image/node_modules/next/dist/compiled/jest-worker/processChild.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1077:15)
    at /home/runner/work/tekblogg/tekblogg/tekblogg-web/node_modules/next-sanity-image/node_modules/next/dist/server/require-hook.js:110:36
    at Module._load (node:internal/modules/cjs/loader:922:27)
    at Module.require (node:internal/modules/cjs/loader:1143:19)
    at require (node:internal/modules/cjs/helpers:121:18)
    at 18743 (/home/runner/work/tekblogg/tekblogg/tekblogg-web/.next/server/pages/_document.js:39:18)
    at __webpack_require__ (/home/runner/work/tekblogg/tekblogg/tekblogg-web/.next/server/webpack-runtime.js:[25](https://github.com/haraldvinje/tekblogg/actions/runs/6292089078/job/17081192160#step:6:26):43)
    at 11692 (/home/runner/work/tekblogg/tekblogg/tekblogg-web/.next/server/chunks/692.js:45:22)
    at __webpack_require__ (/home/runner/work/tekblogg/tekblogg/tekblogg-web/.next/server/webpack-runtime.js:25:43)
    at __webpack_exec__ (/home/runner/work/tekblogg/tekblogg/tekblogg-web/.next/server/pages/_document.js:63:[39](https://github.com/haraldvinje/tekblogg/actions/runs/6292089078/job/17081192160#step:6:40)) {
  type: 'Error',
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/home/runner/work/tekblogg/tekblogg/tekblogg-web/.next/server/pages/_document.js',
    '/home/runner/work/tekblogg/tekblogg/tekblogg-web/node_modules/next-sanity-image/node_modules/next/dist/server/require.js',
    '/home/runner/work/tekblogg/tekblogg/tekblogg-web/node_modules/next-sanity-image/node_modules/next/dist/server/load-components.js',
    '/home/runner/work/tekblogg/tekblogg/tekblogg-web/node_modules/next-sanity-image/node_modules/next/dist/build/utils.js',
    '/home/runner/work/tekblogg/tekblogg/tekblogg-web/node_modules/next-sanity-image/node_modules/next/dist/build/worker.js',
    '/home/runner/work/tekblogg/tekblogg/tekblogg-web/node_modules/next-sanity-image/node_modules/next/dist/compiled/jest-worker/processChild.js'
  ]
}
```

Let me know if more details are needed. Should I submit this as an issue on the [setup-bun](https://github.com/oven-sh/setup-bun) Github Action?

Keep up the good work, bun is awesome 🚀

**Comment by anthonypillot** [comment]
I've exactly the same kind of project like @squishyu ; meaning Docusaurus in V3. Using GitHub Actions, crashing at the build time. Fixed adding `"react-loadable": "npm:@docusaurus/react-loadable@5.5.2"`. My `package.json` dependencies for instance:
```json
"dependencies": {
    "@docusaurus/core": "^3.0.0",
    "@docusaurus/preset-classic": "^3.0.0",
    "@easyops-cn/docusaurus-search-local": "^0.37.4",
    "@mdx-js/react": "^3.0.0",
    "clsx": "^2.0.0",
    "prism-react-renderer": "^2.2.0",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "react-loadable": "npm:@docusaurus/react-loadable@5.5.2"
  }
```

**Comment by niklasravnsborg** [comment]
I also have this problem when running a `docusaurus build` on GitHub Actions via bun. Locally on macOS this works without problems.

**Comment by jakeboone02** [comment]
@squishyu @anthonypillot @niklasravnsborg - have any of you tried this with a more recent Bun version? I was having the same problem a while back but just checked with v1.1.12 and I'm not getting the `react-loadable` error anymore. I wonder if the [workspace fixes in v1.1.9](https://bun.sh/blog/bun-v1.1.9#workspace-fixes) resolved this.

**Comment by Jarred-Sumner** [comment]
yeah I think its likely fixed from that PR, or from the peer dependency fix in v1.1.11

If you're still running into this, please leave a comment and let us know
